### ISSUE TITLE
Proposal: Remove the need for "parent type"

### ISSUE THREAD

**Issue by jeffbaumes** [girder/girder#1338]
"Parent type" crops up in several places - `girder_cli` for one currently requires this flag and for collection download reports "download command only accepts parent-type of folder". But it's not like you really need it - there will never be a Mongo ID that matches both the collection and folder tables. I suppose it is theoretically possible but not practically possible.

On a major version update can we revisit this to perhaps merge both concepts into one model (or a folder model with collection as an extension/subclass)? Collections (1) don't have parents and (2) have extra metadata like quotas, but I believe otherwise behave like folders. The Girder UI could stay the same (i.e. they are still named Collections and Folders), and the parentType parameter could still be supported as a validation check of the expected type, but if absent, it queries a single folder/collection Mongo collection and returns whatever matches. Folders would also no longer need to keep track of whether they are in a folder or collection - they would just have a parent id.

I'm happy to hear the drawbacks of this approach but I feel it's worth considering.

**Comment by jbeezley** [comment]
Agreed.  I would also like to add to the discussion the idea of removing a collection as a distinct model type entirely.  I think you could simply define a collection as any folder without a parent.  As part of this, I would also want to get rid of the collection-like component of the user model to create a simple model hierarchy: file -> item -> folder.    The user page could instead be a list of folders the user has "starred".

I think this would make for a huge reduction in client complexity by eliminating all sorts of validation and special cases that need to occur by detecting the parent type.

**Comment by jeffbaumes** [comment]
Instead of the starred folders idea, perhaps there could be a separation of concerns between a user (account data) and their home folder (a not-deletable folder pointed to by the user). I think the concept of a home folder is useful in Girder just as it is in computer filesystems.

**Comment by zachmullen** [comment]
I like your first proposal:

> you could simply define a collection as any folder without a parent

It doesn't destroy any of our existing domain models that we've been presenting to users, and would simplify our server and client logic. What I find problematic is that it would actually necessitate your second proposal of removing the feature of user data space entirely. I think that's actually a nice feature of Girder.

**Comment by jbeezley** [comment]
I think @jeffbaumes's is decent middle ground.  The important part is that parent for any document should only be of a single type.  By just removing the parent type attribute, you aren't really removing much complexity.  There still need to be special cases every time you fetch a parent because you won't know what type it is before it is returned.

**Comment by zachmullen** [comment]
I don't fully understand that proposal, it touches on collections and folders, but what happens to the user parent type in that case?

**Comment by jeffbaumes** [comment]
There would be no user parent type. Home folders would have no parent, and could have a field that points back to the user, just as a user has a field that points to the home folder. So it this proposal would have something like the following:

```
Folder {
    parent: <null if this is a home folder or collection, Folder ID if non-home folder>,
    homeFolderFor: <null if this is a collection or non-home folder, User ID if home folder>,
    ...
}

User {
    homeFolder: <Folder ID>,
    ...
}
```

**Comment by zachmullen** [comment]
Ok, makes sense. Does the collection model go away completely then?

**Comment by brianhelba** [comment]
@jeffbaumes Why do you have a bidirectional reference, in `Folder.homeFolderFor` and `User.homeFolder`. Assuming the field is indexed, only one should be necessary.

As for where it goes:
* `User.homeFolder` is initially compelling, because it keep the semantics of the `Folder` model as clean as possible. A "home" folder generally shouldn't be treated any differently than others. However, on the special case of home folder deletion, either the operation needs to be prevented, or the `User.homeFolder` needs to be cleaned up; handling this requires a listener to check every single `Folder.remove` operation.
* `Folder.homeFolderFor` doesn't require any extra queries for its standard operations. It should definitely use a [sparse index](https://docs.mongodb.com/manual/core/index-sparse/).

So, assuming `Folder.homeFolderFor` is used, the query to generate the existing list of top-level collections becomes:
`Folder.find({'parent': None, 'homeFolderFor': {'$ne': None}})`.

**Comment by jeffbaumes** [comment]
> Does the collection model go away completely then?

That was my thought, but I admit that I don't know the full implications of that move, like how any collection-specific logic (like quotas) would be handled.

**Comment by manthey** [comment]
It would be pretty straightforward to hook the quotas to folders which have
no parent.

On Thu, May 12, 2016 at 1:44 PM, Jeffrey Baumes <notifications@github.com>
wrote:

> Does the collection model go away completely then?
>
> That was my thought, but I admit that I don't know the full implications
> of that move, like how any collection-specific logic (like quotas) would be
> handled.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/girder/girder/issues/1338#issuecomment-218832394>
>



-- 
David Manthey
R&D Engineer
Kitware Inc.
(518) 881-4439

**Comment by jeffbaumes** [comment]
> @jeffbaumes Why do you have a bidirectional reference, in `Folder.homeFolderFor` and `User.homeFolder`. Assuming the field is indexed, only one should be necessary.

This is very true. I guess I was prematurely optimizing to get the links back and forth without extra queries but you are right it is not needed. Not sure which would be best to keep.

**Comment by zachmullen** [comment]
Actually the bidirectional reference would be a good idea. Remember that we have to traverse up the tree as well as down. For instance, when someone loads the hierarchy widget we traverse all the way to the top of the tree to show them where they are via breadcrumbs.

**Comment by manthey** [comment]
On some API endpoints, we have a small set of possible models that the parent type can be.  We could make it so if the parent type is not specified, that limited list of models is checked before we reject the passed ID.

**Comment by jeffbaumes** [comment]
@manthey that sounds like a simple way to make the APIs easier to use. It could result in 2x the DB queries if parent type is left out, but if that optimization is needed you can always specify the parent type to speed it up.
